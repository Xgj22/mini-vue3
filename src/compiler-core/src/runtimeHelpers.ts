

// 封装常量

export const TO_DISPLAY_STRING = Symbol("toDisplayString")
export const CREATE_ELEMENT_NODE = Symbol("createElementNode")

export const helperMapName = {
    [TO_DISPLAY_STRING] : "toDisplayString",
    [CREATE_ELEMENT_NODE] : "createElementNode"
}