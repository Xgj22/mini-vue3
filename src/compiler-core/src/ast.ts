import { CREATE_ELEMENT_NODE } from "./runtimeHelpers"

export const enum NodeTypes {
    INTERPOLATION,
    SIMPLE_EXPRESSION,
    ELEMENT,
    TEXT,
    ROOT,
    COMPOUND
}

export function createVNodeCall(context,tag,props,children){
    context.helper(CREATE_ELEMENT_NODE)
    return {
        type:NodeTypes.ELEMENT,
        tag,
        props,
        children
    }
}