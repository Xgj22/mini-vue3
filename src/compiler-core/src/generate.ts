import { isString } from "../../shared"
import { NodeTypes } from "./ast"
import { CREATE_ELEMENT_NODE, TO_DISPLAY_STRING, helperMapName } from "./runtimeHelpers"

export function generate(ast){
    const context = createCodegenContext()
    const { push } = context
    
    getFunctionPreamble(ast,context)

    const functionName = 'render'
    const args = ['_ctx','_cache']
    const signature = args.join(', ')

    push(`function ${functionName}(${signature}){`)
    push('return ')
    // 处理节点
    genNode(ast.codegenNode,context)
    push('}')

    return {
        code:context.code
    }
}

function getFunctionPreamble(ast,context){
    const { push } = context
    // toDisplayString 是一直跟着 插值 的
    const VueBinging = "Vue"
    // helpers 应该是基于类型处理，不应该写死在 generate 里
    // const helpers = ['toDisplayString']
    const aliasHelper = (s) => `${helperMapName[s]}:_${helperMapName[s]}`
    // 如果是文本类型，则直接跳过
    if(ast.helpers.length > 0){
        push(`${ast.helpers.map(aliasHelper).join(', ')} = ${VueBinging}`)
    }
    push('\n')
    push('return ')
}

function createCodegenContext() {
    const context = {
        code:'',
        push(source){
            context.code += source
        },
        helper(key){
            return `_${helperMapName[key]}`
        }
    }
    return context
}

function genNode(node,context){
    switch (node.type) {
        case NodeTypes.TEXT:
            genText(node,context)
            break;
        case NodeTypes.ELEMENT:
            genElement(node,context)
            break;
        case NodeTypes.COMPOUND:
            genCompoundExpression(node,context)
            break
        case NodeTypes.INTERPOLATION:
            genInterpolation(node,context)
            break;
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(node,context)
            break;
    }
    
}

function genCompoundExpression(node,context){
    const { children } = node
    const { push } = context
    for(let i = 0;i<children.length;i++){
        const child = children[i]
        if(isString(child)){
            push(child)
        }else{
            genNode(child,context)
        }
    }
}

function genElement(node,context){
    const { tag,props,children } = node
    const { push,helper } = context
    push(`${helper(CREATE_ELEMENT_NODE)}( `)
    genNodeList(getNullable([tag,props,children]),context)
    // genNode(children,context)
    push(')')
}

function getNullable(args){
    return args.map(arg => arg ? arg : 'null')
}

function genNodeList(nodes,context){
    const { push } = context
    for (let i = 0; i < nodes.length; i++) {
        if(isString(nodes[i])){
            push(nodes[i])
        }else{
            genNode(nodes[i],context)
        }

        if(i<nodes.length-1){
            push(',')
        }
    }
}

function genExpression(node,context){
    const { push } = context
    // 处理 ast 树上的数据应该放在 transform 里面
    // push(`_ctx.${node.content}`)
    push(`${node.content}`)
}

function genInterpolation(node,context){
    const { push,helper } = context
    push(`${helper(TO_DISPLAY_STRING)}(`)
    genNode(node.content,context)
    push(')')
}

function genText(node,context){
    const { push } = context
    push(`'${node.content}'`)
}