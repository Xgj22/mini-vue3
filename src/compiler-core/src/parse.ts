import { NodeTypes } from "./ast"

const enum TagType {
    START,
    END
}

export function baseParse(content){
    const context = createParseContext(content)
    return createRoot(parseChildren(context,[]))
}

function parseChildren(context,ancestors) {
    const nodes:any = []

    while(!isEnd(context,ancestors)){
        
        let node
        const s = context.source
        console.log('----------------')
        console.log(ancestors)
        console.log('s==>',s)
        if(s.startsWith('{{')){
            node = parseInterpolation(context)
        }else if(s[0] === '<'){
            if(/[a-z]/i.test(s[1])){
                node = parseElement(context,ancestors)
            }
        }
    
        // context.source 没有匹配到插值和 DOM 标签
        if(!node){
            node = parseText(context)
        }
        nodes.push(node)
    }
    return nodes
}

function isEnd(context,ancestors){
    const s = context.source

    // 当遇到结束标签
    if(s.startsWith('</')){
        for(let i = 0;i<ancestors.length;i++){
            const tag = ancestors[i].tag
            if(s.slice(2,2+tag.length) === tag){
                return true
            }
        }
    }
    
    return !s
}

function parseText(context){

    let endIndex = context.source.length
    
    const endToken = ['<','{{']

    for(let i = 0;i<endToken.length;i++){
        const index = context.source.indexOf(endToken[i])
        
        if(index!==-1 && index < endIndex){
            endIndex = index
        }
    }

    const content = parseTextData(context,endIndex)

    return {
        type:NodeTypes.TEXT,
        content
    }
}

function parseTextData(context,length){
    const content = context.source.slice(0,length)

    // 推进
    advanceBy(context,length)
    return content
}

function parseElement(context,ancestors){

    // 解析 tag
    const element:any = parseTag(context,TagType.START)
    ancestors.push(element)
    element.children = parseChildren(context,ancestors)
    ancestors.pop()
    if(context.source.slice(2,2+element.tag.length)===element.tag){
        // 删除处理完成的代码
        parseTag(context,TagType.END)
    }else{
        console.log(context,element.tag)
        console.log('uuuu',context.source.slice(2,2+element.tag.length))
        throw new Error(`缺少结束标签${element.tag}`)
    }


    return element
}

function parseTag(context,tagType){
    
    const reg = /^<\/?([a-z]*)/

    const match:any = reg.exec(context.source)

    const tag = match[1]

    // 删除处理完成的代码
    advanceBy(context,match[0].length)
    advanceBy(context,1)

    // 当删除多余代码时，无需返回element
    if(tagType==='END') return

    return {
        type:NodeTypes.ELEMENT,
        tag
    }
}

function parseInterpolation(context) {

    const openDelimiter = '{{'
    const closeDelimiter = '}}'

    const closeIndex = context.source.indexOf(closeDelimiter,closeDelimiter.length)

    advanceBy(context,openDelimiter.length)

    const rawContentLength = closeIndex - openDelimiter.length
    
    const content = parseTextData(context,rawContentLength).trim()

    advanceBy(context,closeDelimiter.length)

    return {
        type:NodeTypes.INTERPOLATION,
        content:{
            type:NodeTypes.SIMPLE_EXPRESSION,
            content:content
        }
    }
}

// 推进
function advanceBy(context,length){
    context.source = context.source.slice(length)
}

function createRoot(children){
    return {
        children
    }
}

function createParseContext(content) {
    return {
        source:content
    }
}