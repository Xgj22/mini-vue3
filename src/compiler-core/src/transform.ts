import { NodeTypes } from "./ast"
import { TO_DISPLAY_STRING } from "./runtimeHelpers"

export function transform(root,options = {}){
    const context = createTransformcontext(root,options)
    // 1. 遍历 - 深度优先搜索
    traverseNode(root,context)

    // 2. 修改 text content

    // 语义化 codegenNode，ast 树的入口
    createRootCodegen(root)
    root.helpers =[...context.helpers.keys()]
}

function createRootCodegen(root){
    const child = root.children[0]
    if(child.type===NodeTypes.ELEMENT){
        root.codegenNode = child.codegenNode
    }else{
        root.codegenNode = root.children[0]
    }
}

function createTransformcontext(node,options){
    const context = {
        node,
        nodeTransform:options.nodeTransform || [],
        helpers:new Map(),
        helper(key){
            context.helpers.set(key,1)
        }
    }
    return context
}

function traverseNode(node:any,context){

    const nodeTransform = context.nodeTransform
    const exitFns:any  = []

    for(let i = 0;i<nodeTransform.length;i++){
        const transform = nodeTransform[i]
        // 先调用 后执行
        const onExit = transform(node,context)
        if(onExit) exitFns.push(onExit)
    }

    switch (node.type) {
        case NodeTypes.INTERPOLATION:
            context.helper(TO_DISPLAY_STRING)
            break;
        // 只有当 节点是根节点或 标签节点的时候 才需要去遍历子节点
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            traverseChildren(node,context)
            break;
        default:
            break;
    }

    let i = exitFns.length
    while(i--){
        exitFns[i]()
    }
}

function traverseChildren(node,context){
    const children = node.children

    for(let i = 0;i<children.length;i++){
        traverseNode(children[i],context)
    }
    
}