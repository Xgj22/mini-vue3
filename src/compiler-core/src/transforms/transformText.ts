import { NodeTypes } from "../ast";


export function transfromText(node){
    if(node.type===NodeTypes.ELEMENT){
        
        return () => {
            const { children } = node

            function isText(node){
                return node.type === NodeTypes.INTERPOLATION || node.type === NodeTypes.TEXT
            }

            let currentContainer
            // 设置一个新的类型 compound ，包裹 text 和 插值
            for(let i = 0;i<children.length;i++){
                const child = children[i]
                if(isText(child)){
                    for(let j = i+1;j<children.length;j++){
                        const next = children[j]
                        if(isText(next)){
                            if(!currentContainer){
                                currentContainer = children[i] = {
                                    type:NodeTypes.COMPOUND,
                                    children:[child]
                                }
                            }
                            currentContainer.children.push(' + ')
                            currentContainer.children.push(next)
                            children.splice(j,1)
                            j--
                        }else{
                            currentContainer = undefined
                            break
                        }
                    }
                }
            }
        }
    }
}