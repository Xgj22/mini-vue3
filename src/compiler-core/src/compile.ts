import { generate } from "./generate";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transfromText } from "./transforms/transformText";

export function baseCompile(template){
    const ast:any = baseParse(template)
    transform(ast,{
        nodeTransform:[transformExpression,transformElement,transfromText] // 
    })
    return generate(ast)

}