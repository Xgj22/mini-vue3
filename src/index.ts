// mini-vue 出口
export * from './runtime-dom'

import { baseCompile } from './compiler-core/src/compile'
import { registerRuntimeCompiler } from './runtime-core/component'
import * as runtimeDom from './runtime-dom'

function compileToFunction(template){
    const { code } = baseCompile(template)
    console.log(code)
    const render = new Function("Vue",code)(runtimeDom)
    console.log(render)
    return render

}

registerRuntimeCompiler(compileToFunction)