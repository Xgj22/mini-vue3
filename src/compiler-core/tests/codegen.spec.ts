import { generate } from "../src/generate"
import { baseParse } from "../src/parse"
import { transform } from "../src/transform"
import { transformElement } from "../src/transforms/transformElement"
import { transformExpression } from "../src/transforms/transformExpression"
import { transfromText } from "../src/transforms/transformText"


describe("codegen",() => {
    it("string",() => {
        const ast = baseParse("hi")

        transform(ast)
        const { code } = generate(ast)

        // 快照
        // 1. 抓 bug
        // 2. 有意（）
        expect(code).toMatchSnapshot()
    })

    it("iterpolation",() => {
        const ast = baseParse("{{message}}")

        transform(ast,{
            nodeTransform:[transformExpression]
        })
        const { code } = generate(ast)

        expect(code).toMatchSnapshot()
    })

    it("element",() => {
        const ast:any = baseParse("<div>hi,{{message}}</div>")

        transform(ast,{
            // nodeTransform:[transformElement,transfromText]
            nodeTransform:[transformExpression,transformElement,transfromText] // 
        })
        console.log(ast.codegenNode.children[0])
        const { code } = generate(ast)

        expect(code).toMatchSnapshot()
    })
})

