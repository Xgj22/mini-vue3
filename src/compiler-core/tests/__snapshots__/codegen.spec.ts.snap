// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"toDisplayString:_toDisplayString, createElementNode:_createElementNode = Vue
return function render(_ctx, _cache){return _createElementNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;

exports[`codegen iterpolation 1`] = `
"toDisplayString:_toDisplayString = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx, _cache){return 'hi'}"
`;
